File: ios-demo/BankLiteDemo/Classes/CoreFoundation/ColorPalette.swift
//
//  ColorPalette.swift
//  BankLiteDemo
//
//  Created by Алмазов Иван Александрович on 04.04.2022.
//

import UIKit
import SwiftUI

public enum ColorPaletteStyle: String, Codable {
    case light
    case dark
}

// MARK: - ColorPalette

public class ColorPalette: Codable {
        
    public var style: ColorPaletteStyle
    public var disabledAlpha: CGFloat
     
    public var textPrimary: AColor
    public var textSecondary: AColor
    public var textAccent: AColor
    public var textError: AColor
    public var textAttention: AColor
    public var textExtraSuccess: AColor
    public var textStaticWhite: AColor
    public var textStaticBlack: AColor
    public var elementsAccent: AColor
    public var elementsSecondary: AColor
    public var elementsSpecial: AColor
    public var elementsError: AColor
    public var elementsAttention: AColor
    public var elementsGreen: AColor
    public var elementsStaticWhite: AColor
    public var elementsExtraGrey: AColor
    public var backgroundAdditional: AColor
    public var backgroundBasic: AColor
    public var backgroundSelected: AColor
    public var backgroundAccent: AColor
    public var backgroundExtra: AColor
    public var backgroundError: AColor
    public var backgroundAttention: AColor
    public var backgroundSuccess: AColor
    public var backgroundCards: AColor
    public var airBarMain: AColor
    public var airBarMainLight: AColor
    public var airBarNewYear: AColor
    public var airBarLavanda: AColor
    public var airBarDark: AColor
    public var airBarLight: AColor
    public var widgetBackground: AColor
    public var widgetSection: AColor
    public var widgetIcon: AColor
    public var cloudIcon: AColor
    

    public init(
        style: ColorPaletteStyle,
        disabledAlpha: CGFloat,
        textPrimary: AColor,
        textSecondary: AColor,
        textAccent: AColor,
        textError: AColor,
        textAttention: AColor,
        textExtraSuccess: AColor,
        textStaticWhite: AColor,
        textStaticBlack: AColor,
        elementsAccent: AColor,
        elementsSecondary: AColor,
        elementsSpecial: AColor,
        elementsError: AColor,
        elementsAttention: AColor,
        elementsGreen: AColor,
        elementsStaticWhite: AColor,
        elementsExtraGrey: AColor,
        backgroundAdditional: AColor,
        backgroundBasic: AColor,
        backgroundSelected: AColor,
        backgroundAccent: AColor,
        backgroundExtra: AColor,
        backgroundError: AColor,
        backgroundAttention: AColor,
        backgroundSuccess: AColor,
        backgroundCards: AColor,
        airBarMain: AColor,
        airBarMainLight: AColor,
        airBarNewYear: AColor,
        airBarLavanda: AColor,
        airBarDark: AColor,
        airBarLight: AColor,
        widgetBackground: AColor,
        widgetSection: AColor,
        widgetIcon: AColor,
        cloudIcon: AColor
    ) {
        self.style = style
        self.disabledAlpha = disabledAlpha
        
        self.disabledAlpha = disabledAlpha
        self.textPrimary = textPrimary
        self.textSecondary = textSecondary
        self.textAccent = textAccent
        self.textError = textError
        self.textAttention = textAttention
        self.textExtraSuccess = textExtraSuccess
        self.textStaticWhite = textStaticWhite
        self.textStaticBlack = textStaticBlack
        self.elementsAccent = elementsAccent
        self.elementsSecondary = elementsSecondary
        self.elementsSpecial = elementsSpecial
        self.elementsError = elementsError
        self.elementsAttention = elementsAttention
        self.elementsGreen = elementsGreen
        self.elementsStaticWhite = elementsStaticWhite
        self.elementsExtraGrey = elementsExtraGrey
        self.backgroundAdditional = backgroundAdditional
        self.backgroundBasic = backgroundBasic
        self.backgroundSelected = backgroundSelected
        self.backgroundAccent = backgroundAccent
        self.backgroundExtra = backgroundExtra
        self.backgroundError = backgroundError
        self.backgroundAttention = backgroundAttention
        self.backgroundSuccess = backgroundSuccess
        self.backgroundCards = backgroundCards
        self.airBarMain = airBarMain
        self.airBarMainLight = airBarMainLight
        self.airBarNewYear = airBarNewYear
        self.airBarLavanda = airBarLavanda
        self.airBarDark = airBarDark
        self.airBarLight = airBarLight
        self.widgetBackground = widgetBackground
        self.widgetSection = widgetSection
        self.widgetIcon = widgetIcon
        self.cloudIcon = cloudIcon
        
    }
}
 
// MARK: - Light

extension ColorPalette {

    public static var light: ColorPalette {
        return ColorPalette(style: .light,
                            disabledAlpha: 0.6,
                            textPrimary: AColor(hexString: "#FF222222"),
                            textSecondary: AColor(hexString: "#FF818A97"),
                            textAccent: AColor(hexString: "#FF3A83F1"),
                            textError: AColor(hexString: "#FFF5443A"),
                            textAttention: AColor(hexString: "#FFFB9130"),
                            textExtraSuccess: AColor(hexString: "#FF45BF78"),
                            textStaticWhite: AColor(hexString: "#FFFFFFFF"),
                            textStaticBlack: AColor(hexString: "#FF222222"),
                            elementsAccent: AColor(hexString: "#FF3A83F1"),
                            elementsSecondary: AColor(hexString: "#FF99A0AB"),
                            elementsSpecial: AColor(hexString: "#FFDDDFE3"),
                            elementsError: AColor(hexString: "#FFF5443A"),
                            elementsAttention: AColor(hexString: "#FFFB9130"),
                            elementsGreen: AColor(hexString: "#FF45BF78"),
                            elementsStaticWhite: AColor(hexString: "#FFFFFFFF"),
                            elementsExtraGrey: AColor(hexString: "#FF546173"),
                            backgroundAdditional: AColor(hexString: "#FFFFFFFF"),
                            backgroundBasic: AColor(hexString: "#FFF0F2F5"),
                            backgroundSelected: AColor(hexString: "#FFEBF3FE"),
                            backgroundAccent: AColor(hexString: "#FF3A83F1"),
                            backgroundExtra: AColor(hexString: "#FF546173"),
                            backgroundError: AColor(hexString: "#FFFEECEB"),
                            backgroundAttention: AColor(hexString: "#FFFEF4EB"),
                            backgroundSuccess: AColor(hexString: "#FFEDF8F2"),
                            backgroundCards: AColor(hexString: "#FFFFFFFF"),
                            airBarMain: AColor(hexString: "FF3A4A61"),
                            airBarMainLight: AColor(hexString: "#FF7E8EA8"),
                            airBarNewYear: AColor(hexString: "#FF33837D"),
                            airBarLavanda: AColor(hexString: "#FF71698C"),
                            airBarDark: AColor(hexString: "#FF00051E"),
                            airBarLight: AColor(hexString: "#FFF4FAFF"),
                            widgetBackground: AColor(hexString: "#FFF0F2F5"),
                            widgetSection: AColor(hexString: "#FFFFFFFF"),
                            widgetIcon: AColor(hexString: "#FFF0F2F5"),
                            cloudIcon: AColor(hexString: "#FFE0F2FF"))
    }

}

// MARK: - Dark
 
extension ColorPalette {

    public static var dark: ColorPalette {
        return ColorPalette(style: .dark,
                            disabledAlpha: 0.6,
                            textPrimary: AColor(hexString: "#FFE4E8EE"),
                            textSecondary: AColor(hexString: "#FFA5ACB6"),
                            textAccent: AColor(hexString: "#FF5594F1"),
                            textError: AColor(hexString: "#FF5594F1"),
                            textAttention: AColor(hexString: "#FFF09B4C"),
                            textExtraSuccess: AColor(hexString: "#FF60C78B"),
                            textStaticWhite: AColor(hexString: "#FFFFFFFF"),
                            textStaticBlack: AColor(hexString: "#FF222222"),
                            elementsAccent: AColor(hexString: "#FF5594F1"),
                            elementsSecondary: AColor(hexString: "#FF99A0AB"),
                            elementsSpecial: AColor(hexString: "#FF4E555F"),
                            elementsError: AColor(hexString: "#FFF76A64"),
                            elementsAttention: AColor(hexString: "#FFF09B4C"),
                            elementsGreen: AColor(hexString: "#FF60C78B"),
                            elementsStaticWhite: AColor(hexString: "#FFFFFFFF"),
                            elementsExtraGrey: AColor(hexString: "#FFADB3BC"),
                            backgroundAdditional: AColor(hexString: "#292F38"),
                            backgroundBasic: AColor(hexString: "#FF1C2026"),
                            backgroundSelected: AColor(hexString: "#FF363E4A"),
                            backgroundAccent: AColor(hexString: "#FF5594F1"),
                            backgroundExtra: AColor(hexString: "#FFADB3BC"),
                            backgroundError: AColor(hexString: "#FF3F2424"),
                            backgroundAttention: AColor(hexString: "#FF3C2F20"),
                            backgroundSuccess: AColor(hexString: "#FF22392D"),
                            backgroundCards: AColor(hexString: "#FF303641"),
                            airBarMain: AColor(hexString: "FF3A4A61"),
                            airBarMainLight: AColor(hexString: "#FF7E8EA8"),
                            airBarNewYear: AColor(hexString: "#FF33837D"),
                            airBarLavanda: AColor(hexString: "#FF71698C"),
                            airBarDark: AColor(hexString: "#FF00051E"),
                            airBarLight: AColor(hexString: "#FFF4FAFF"),
                            widgetBackground: AColor(hexString: "#FF050506"),
                            widgetSection: AColor(hexString: "#FF1C2026"),
                            widgetIcon: AColor(hexString: "#FF363E49"),
                            cloudIcon: AColor(hexString: "#FF343C48"))
    }

}

// MARK: - Color

public struct AColor: Codable {
    public let red: CGFloat
    public let green: CGFloat
    public let blue: CGFloat
    public let alpha: CGFloat
    
    public var uiColor: UIColor {
        return UIColor(red: red, green: green, blue: blue, alpha: alpha)
    }

    public var cgColor: CGColor {
        return uiColor.cgColor
    }
    
    public static var clear: AColor {
        return AColor(red: 0, green: 0, blue: 0, alpha: 0.0)
    }
    
    public init(red: CGFloat, green: CGFloat, blue: CGFloat, alpha: CGFloat) {
        self.red = red
        self.green = green
        self.blue = blue
        self.alpha = alpha
    }
    
    public init(red: Int, green: Int, blue: Int) {
        self.init(red: CGFloat(red) / 255.0,
                  green: CGFloat(green) / 255.0,
                  blue: CGFloat(blue) / 255.0,
                  alpha: 1.0)
    }
    
    public init(hexString: String) {
        let hex = hexString.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int = UInt64()
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 3: // RGB (12-bit)
            (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
        case 6: // RGB (24-bit)
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB (32-bit)
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            (a, r, g, b) = (255, 0, 0, 0)
        }
        
        self.init(red: CGFloat(r) / 255, green: CGFloat(g) / 255, blue: CGFloat(b) / 255, alpha: CGFloat(a) / 255)
    }
    
    public init(color: UIColor) {
        let colorTo = CGColorSpace(name: CGColorSpace.sRGB)!
        let cgColorInRGB = color.cgColor.converted(
            to: colorTo,
            intent: .defaultIntent,
            options: nil)!
        
        let colorRef = cgColorInRGB.components
        let red = colorRef?[0] ?? 0
        let green = colorRef?[1] ?? 0
        let blue = ((colorRef?.count ?? 0) > 2 ? colorRef?[2] : green) ?? 0
        let alpha = color.cgColor.alpha
        self.init(red: red, green: green, blue: blue, alpha: alpha)
    }
    
    public func hexString(withAplha: Bool = true) -> String {
        var colorString = withAplha ? String(format: "#%02lX", lroundf(Float(alpha * 255))) : "#"

        colorString += String(
            format: "%02lX%02lX%02lX",
            lroundf(Float(red * 255)),
            lroundf(Float(green * 255)),
            lroundf(Float(blue * 255))
        )
    
        return colorString
    }
    
    public func withAlpha(_ alpha: CGFloat) -> AColor {
        return AColor(red: red, green: green, blue: blue, alpha: alpha)
    }
    
}

@available(iOS 13.0.0, *)
extension AColor {
    
    public var swiftUIColor: SwiftUI.Color {
        return SwiftUI.Color(.sRGB, red: Double(red), green: Double(green), blue: Double(blue), opacity: Double(alpha))
    }
    
}

extension UIColor {
    
    public var color: AColor {
        return AColor(color: self)
    }
    
}

// swiftlint:enable all

File: flutter-demo/lib/theme/color_palette.dart
import 'package:bank_lite/theme/color_wrapper.dart';

class ColorPalette {
  final ColorWrapper textPrimary;
  final ColorWrapper textSecondary;
  final ColorWrapper textAccent;
  final ColorWrapper textError;
  final ColorWrapper textAttention;
  final ColorWrapper textExtraSuccess;
  final ColorWrapper textStaticWhite;
  final ColorWrapper textStaticBlack;
  final ColorWrapper elementsAccent;
  final ColorWrapper elementsSecondary;
  final ColorWrapper elementsSpecial;
  final ColorWrapper elementsError;
  final ColorWrapper elementsAttention;
  final ColorWrapper elementsGreen;
  final ColorWrapper elementsStaticWhite;
  final ColorWrapper elementsExtraGrey;
  final ColorWrapper backgroundAdditional;
  final ColorWrapper backgroundBasic;
  final ColorWrapper backgroundSelected;
  final ColorWrapper backgroundAccent;
  final ColorWrapper backgroundExtra;
  final ColorWrapper backgroundError;
  final ColorWrapper backgroundAttention;
  final ColorWrapper backgroundSuccess;
  final ColorWrapper backgroundCards;
  final ColorWrapper airBarMain;
  final ColorWrapper airBarMainLight;
  final ColorWrapper airBarNewYear;
  final ColorWrapper airBarLavanda;
  final ColorWrapper airBarDark;
  final ColorWrapper airBarLight;
  final ColorWrapper widgetBackground;
  final ColorWrapper widgetSection;
  final ColorWrapper widgetIcon;
  final ColorWrapper cloudIcon;

  const ColorPalette({
    required this.textPrimary,
    required this.textSecondary,
    required this.textAccent,
    required this.textError,
    required this.textAttention,
    required this.textExtraSuccess,
    required this.textStaticWhite,
    required this.textStaticBlack,
    required this.elementsAccent,
    required this.elementsSecondary,
    required this.elementsSpecial,
    required this.elementsError,
    required this.elementsAttention,
    required this.elementsGreen,
    required this.elementsStaticWhite,
    required this.elementsExtraGrey,
    required this.backgroundAdditional,
    required this.backgroundBasic,
    required this.backgroundSelected,
    required this.backgroundAccent,
    required this.backgroundExtra,
    required this.backgroundError,
    required this.backgroundAttention,
    required this.backgroundSuccess,
    required this.backgroundCards,
    required this.airBarMain,
    required this.airBarMainLight,
    required this.airBarNewYear,
    required this.airBarLavanda,
    required this.airBarDark,
    required this.airBarLight,
    required this.widgetBackground,
    required this.widgetSection,
    required this.widgetIcon,
    required this.cloudIcon,
  });
}

const lightPalette = ColorPalette(
  textPrimary: ColorWrapper(0xFF222222),
  textSecondary: ColorWrapper(0xFF818A97),
  textAccent: ColorWrapper(0xFF3A83F1),
  textError: ColorWrapper(0xFFF5443A),
  textAttention: ColorWrapper(0xFFFB9130),
  textExtraSuccess: ColorWrapper(0xFF45BF78),
  textStaticWhite: ColorWrapper(0xFFFFFFFF),
  textStaticBlack: ColorWrapper(0xFF222222),
  elementsAccent: ColorWrapper(0xFF3A83F1),
  elementsSecondary: ColorWrapper(0xFF99A0AB),
  elementsSpecial: ColorWrapper(0xFFDDDFE3),
  elementsError: ColorWrapper(0xFFF5443A),
  elementsAttention: ColorWrapper(0xFFFB9130),
  elementsGreen: ColorWrapper(0xFF45BF78),
  elementsStaticWhite: ColorWrapper(0xFFFFFFFF),
  elementsExtraGrey: ColorWrapper(0xFF546173),
  backgroundAdditional: ColorWrapper(0xFFFFFFFF),
  backgroundBasic: ColorWrapper(0xFFF0F2F5),
  backgroundSelected: ColorWrapper(0xFFEBF3FE),
  backgroundAccent: ColorWrapper(0xFF3A83F1),
  backgroundExtra: ColorWrapper(0xFF546173),
  backgroundError: ColorWrapper(0xFFFEECEB),
  backgroundAttention: ColorWrapper(0xFFFEF4EB),
  backgroundSuccess: ColorWrapper(0xFFEDF8F2),
  backgroundCards: ColorWrapper(0xFFFFFFFF),
  airBarMain: ColorWrapper(0x0FF3A4A61),
  airBarMainLight: ColorWrapper(0xFF7E8EA8),
  airBarNewYear: ColorWrapper(0xFF33837D),
  airBarLavanda: ColorWrapper(0xFF71698C),
  airBarDark: ColorWrapper(0xFF00051E),
  airBarLight: ColorWrapper(0xFFF4FAFF),
  widgetBackground: ColorWrapper(0xFFF0F2F5),
  widgetSection: ColorWrapper(0xFFFFFFFF),
  widgetIcon: ColorWrapper(0xFFF0F2F5),
  cloudIcon: ColorWrapper(0xFFE0F2FF),
);

const darkPalette = ColorPalette(
  textPrimary: ColorWrapper(0xFFE4E8EE),
  textSecondary: ColorWrapper(0xFFA5ACB6),
  textAccent: ColorWrapper(0xFF5594F1),
  textError: ColorWrapper(0xFFF76A64),
  textAttention: ColorWrapper(0xFFF09B4C),
  textExtraSuccess: ColorWrapper(0xFF60C78B),
  textStaticWhite: ColorWrapper(0xFFFFFFFF),
  textStaticBlack: ColorWrapper(0xFF222222),
  elementsAccent: ColorWrapper(0xFF5594F1),
  elementsSecondary: ColorWrapper(0xFF99A0AB),
  elementsSpecial: ColorWrapper(0xFF4E555F),
  elementsError: ColorWrapper(0xFFF76A64),
  elementsAttention: ColorWrapper(0xFFF09B4C),
  elementsGreen: ColorWrapper(0xFF60C78B),
  elementsStaticWhite: ColorWrapper(0xFFFFFFFF),
  elementsExtraGrey: ColorWrapper(0xFFADB3BC),
  backgroundAdditional: ColorWrapper(0xFF292F38),
  backgroundBasic: ColorWrapper(0xFF1C2026),
  backgroundSelected: ColorWrapper(0xFF363E4A),
  backgroundAccent: ColorWrapper(0xFF5594F1),
  backgroundExtra: ColorWrapper(0xFFADB3BC),
  backgroundError: ColorWrapper(0xFF3F2424),
  backgroundAttention: ColorWrapper(0xFF3C2F20),
  backgroundSuccess: ColorWrapper(0xFF22392D),
  backgroundCards: ColorWrapper(0xFF303641),
  airBarMain: ColorWrapper(0xFF3A4A61),
  airBarMainLight: ColorWrapper(0xFF7E8EA8),
  airBarNewYear: ColorWrapper(0xFF33837D),
  airBarLavanda: ColorWrapper(0xFF71698C),
  airBarDark: ColorWrapper(0xFF00051E),
  airBarLight: ColorWrapper(0xFFF4FAFF),
  widgetBackground: ColorWrapper(0xFF050506),
  widgetSection: ColorWrapper(0xFF1C2026),
  widgetIcon: ColorWrapper(0xFF363E49),
  cloudIcon: ColorWrapper(0xFF343C48),
);
